Auto Login to ITSI when coming from an activity

rajadain opened the pull request:
## Overview

When a user initially accesses a URL from within an activity in the ITSI portal, we can have the portal append a flag to the URL indicating that they are coming from the portal. Currently, we assume this flag to be a ~~`?initialize_itsi=true`~~ `?itsi_embed=true` query parameter. This is configurable in the ITSI ~~`url_flag`~~ `embed_flag` setting in `base.py`.

We intercept all requests to check if this query parameter exists. If it does not, we do nothing. ~~If it does, we see if the user is already logged in, and if so, do they have an ITSI account. If they do, we do nothing.~~ Otherwise, we attempt to log them in automatically via the portal. Since they are already logged in, the redirection will be transparent and they will be returned to the initial request path after the oAuth dance is done. If there is any error during this, they will instead be redirected to `/error/itsi`.

In case they don't yet have an MMW account corresponding to their ITSI account, they will be prompted to sign up, after which they will be returned to the initial request path.

## Testing Instructions

 1. Go to the home page of the app. Log out.
 2. Go to ~~`/?initialize_itsi=true`~~ [`/?itsi_embed=true`](http://localhost:8000/?itsi_embed=true). An attempt should be made to automatically log you in to ITSI. If it is successful, you should be returned to ~~`/?initialize_itsi=true`~~ [`/`](http://localhost:8000/).
 3. Go to a public project, for example `/project/1/scenario/2`. Log out of your ITSI account. Go to the same project, but this time add the parameter: ~~`/project/1/scenario/2?initialize_itsi=true`~~ `/project/1/scenario/2?itsi_embed=true`. You should be automatically logged in and returned to the original URL.
 4. Try doing that with a private project. You will first be logged in to ITSI, then you should get a `404`.
 5. Delete your ITSI account from the database:

    ```sql
    DELETE FROM user_itsiuser WHERE user_id IN (
        SELECT id FROM auth_user WHERE username = 'tstudent2'
    );
    DELETE FROM auth_user WHERE username = 'tstudent2';
    ```

  Then try accessing the public project. You should be prompted to sign up as a new user, after which you should be returned to the original project.
 6. When logged in as one ITSI user in the system, `tstudent2`, log in to MMW. Then go back to ITSI and log out, and log in as another ITSI user `mmwterence`. Then come back to MMW and add the [`/?itsi_embed=true`](http://localhost:8000/?itsi_embed=true) flag. You should be logged in as the current ITSI user, namely `mmwterence`.

## Questions

 * ~~Currently, after logging in with ITSI, we **always** set the `initialize_itsi` flag. Should we always do this, or only when they are coming in from the ITSI portal?~~
 * ~~Currently, we only check if they are logged in with an ITSI account. Should we try and check to see if they are logged in as the same user as they are in ITSI? For example, a teacher might have a dummy student account to check what things look like from that perspective, and may forget to log out. They may be logged in with the dummy student account in MMW even if they are logged in as themselves in ITSI.  
  However, that would require some refactoring, since currently we use the fact that the logged in user is an ITSI user to escape the loop of always checking against the portal, since the `intialize_itsi` flag is set even when returning after logging in, to facilitate behavior expected in #560 and #561.~~

Connects #562

rajadain commented on src/mmw/mmw/settings/base.py:
```diff
@@ -244,6 +244,7 @@ def get_env_setting(setting):
     'django.contrib.sessions.middleware.SessionMiddleware',
     'django.middleware.csrf.CsrfViewMiddleware',
     'django.contrib.auth.middleware.AuthenticationMiddleware',
+    'apps.user.middleware.ItsiAuthenticationMiddleware',
```
Is this the right place to put the new middleware? I also see that we have `mmw.middleware.BypassMiddleware`, and I could have added my `ItsiAuthenticationMiddleware` class there, but felt like this was more cohesive since all the other ITSI stuff is contained within `apps.user`.

lewfish replied:
If a middleware returns a redirect, then subsequent middlewares will not run, according to the Django docs. So it's safer to put it at the end of the list.

rajadain replied:
Agreed.

rajadain commented on src/mmw/mmw/settings/base.py:
```diff
@@ -358,6 +359,9 @@ def get_env_setting(setting):
     'authorize_url': 'auth/concord_id/authorize',
     'access_token_url': 'auth/concord_id/access_token',
     'user_json_url': 'auth/concord_id/user.json',
+
```
This blank line added to differentiate the ITSI Portal specific settings, and MMW specific settings (which follow).

rajadain commented on src/mmw/apps/user/views.py:
```diff
@@ -81,10 +82,14 @@ def logout(request):
 
 def itsi_login(request):
-    redirect_uri = request.build_absolute_uri(reverse('itsi_auth'))
+    redirect_uri = '{0}?next={1}'.format(
```
Do we prefer `.format` or `%` for string interpolation?

kdeloach replied:
Either is fine, I prefer `.format` slightly more. Those Python guidelines are not authored by anyone on the Civic Apps team and could probably be updated.

rajadain replied:
I need `.format` for auto parsing keys from a dictionary here: https://github.com/WikiWatershed/model-my-watershed/pull/578/files#diff-76f6d7d05a4b988f9b9af9034c53a081R119 so I decided to be consistent and just use it everywhere. Reading up on recent Python guidelines, it seems like there is a gradual movement away from `%`

rajadain commented on src/mmw/apps/user/views.py:
```diff
@@ -108,12 +113,18 @@ def itsi_auth(request):
     if user is not None and user.is_active:
         auth_login(request, user)
-        return redirect('/')
+        redirect_uri = '{0}?{1}=true'.format(
```
Is it okay to redirect from strings or should we use something more explicitly defined like `request.build_absolute_uri(reverse('itsi_login'))` or something?

rajadain commented on src/mmw/apps/user/middleware.py:
```diff
@@ -0,0 +1,39 @@
+
+        # Flag is set and the user is not already logged in.
+        # Redirect them to ITSI LOGIN URL, with the current URL to return to
+        return redirect('{0}?next={1}'.format(LOGIN_URL, request.path))
```
Is it okay to redirect from strings or should we use something more explicitly defined like `request.build_absolute_uri(reverse(...))` or something?

rajadain commented on src/mmw/js/src/user/views.js:
```diff
@@ -355,7 +355,7 @@ var ItsiSignUpModalView = ModalBaseView.extend({
     // Override to provide custom success handler
     primaryAction: function() {
-        var formData = this.$el.find('form').serialize();
+        var formData = this.model.toJSON();
```
We do this since we set the `next` variable on the model, not in the form.

mmcfarland commented:
Functionality is working really well, and I like the approach.  Going to spend some time looking at the code, in the meantime:

* No, I think we should only be setting the itsi flag on the QS when it was provided in the initial request.  Seems like it'll be good for the app to know whether it is in an Interactive or not, and that flag seems like the clearest indication.
* I think if the user is logged in, and we've received the itsi flag, we should essentially re-login to ensure that we reflect the current session in ITSI.  Unsure based on your comment if that would require much refactoring.

mmcfarland commented:
Also, should we think about adding a session variable (or similar) to indicate coming from an ITSI session, rather than preserve the querystring after parsing and doing the login dance?

rajadain commented:
Ah, session variables! I'd forgotten about those. You're right, we should set that and not have it in the URL all the time. I had thought about using cookies but remembered that they sometimes behave strangely when in iframes.

The issue with resetting the flag in the URL is for the use case when the user lands on a project page, then we do the oAuth dance, and then redirect them back to the same page, which we must now clone into their account. Since cloning so far expects to use the same flag, it must be present in the URL. However, when we see the flag we will try to log them in again (to make sure they're using the current ITSI account), and thus be stuck in a loop.

The way out of this loop seems to be to make a separate flag / endpoint for project cloning. I'll meditate on that and add a solution here and comments on #560.

hectcastro commented:
@azavea-bot rebuild

rajadain commented:
I have simplified the PR according to received feedback and updated the main description and testing instructions. This is ready to be looked at again.

rajadain commented on src/mmw/js/src/user/views.js:
```diff
@@ -335,11 +343,6 @@ var ItsiSignUpModalView = ModalBaseView.extend({
         this.ui.username.focus();
     },
 
-    onModalHidden: function() {
```
Removed this since we're handling the re-routing in the end of `handleSuccess`

lewfish commented on src/mmw/apps/user/views.py:
```diff
@@ -84,7 +84,10 @@ def logout(request):
-    redirect_uri = request.build_absolute_uri(reverse('itsi_auth'))
+    redirect_uri = '{0}?next={1}'.format(
+        request.build_absolute_uri(reverse(itsi_auth)),
+        request.GET.get('next', '/')
```
This expression is repeated several times so you could pull it into a variable.

lewfish commented on src/mmw/js/src/user/controllers.js:
```diff
@@ -12,13 +13,14 @@ var SignUpController = {
                 first_name: first_name,
-                last_name: last_name
+                last_name: last_name,
+                next: queryParams ? utils.parseQueryString(queryParams).next : null
```
I'm not sure, but maybe the null should be a '/'.

rajadain replied:
Hmm at first I was thinking that the back-end always has to redirect somewhere, so we default to `/` if no `next` is specified. The front-end on the other hand has the third option of not redirecting at all, which is what the current implementation does here: https://github.com/WikiWatershed/model-my-watershed/pull/578/files#diff-bbc329caa6aad1d9e2bf745db301390fR381. However, now that I think about it, I think we do always want to redirect away from the sign-up URL, even if nothing is specified, so you're right this should default to `/`. Good catch.

lewfish commented:
+1, tested. Nicely done.

rajadain commented on src/mmw/apps/user/middleware.py:
```diff
@@ -0,0 +1,34 @@
+
+        # Flag is set. Assume user is not logged in. Set session flag and
+        # Redirect them to ITSI LOGIN URL, with the current URL to return to
+        request.session[EMBED_FLAG] = True
```
It's not used as part of this PR, but we might need it for #561 or #563 or #564.

rajadain commented:
Moved middleware to end of chain, set client to default `next` to `/` so it always redirects even when `next` is not specified.
